### ISSUE TITLE
ksh u+m does not add syntax errors to history, corrupts history

### ISSUE THREAD

**Issue by posguy99** [ksh93/ksh#209]
Re the thread at https://groups.google.com/g/korn-shell/c/IGAHm-5UF4o/m/PZk3mqxMAQAJ ...

Something is definitely going on with history in u+m.  Example:

```
Version AJMv 93u+m/1.0.0-alpha+c1986c4e 2021-03-06
[11:57 AM][ttys007 +1][~]
[1] mbp13 $ for i in * do echo $i ; done                                       
-ksh: syntax error: `done' unexpected
[11:58 AM][ttys007 +1][~]
[252639022] mbp13 $ 
```

Uh, WHAT?

```
[11:59 AM][ttys007 +1][~]
[252639022] mbp13 $ history
252639007	?for i in * do echo $i ; done
252639008	?for i in * do echo $i ; done
252639009	?for i in * do echo $i ; done
252639010	?for i in * do echo $i ; done
252639011	?for i in * do echo $i ; done
252639012	?for i in * do echo $i ; done
252639013	?for i in * do echo $i ; done
252639014	?for i in * do echo $i ; done
252639015	?for i in * do echo $i ; done
252639016	?for i in * do echo $i ; done
252639017	?for i in * do echo $i ; done
252639018	?for i in * do echo $i ; done
252639019	?for i in * do echo $i ; done
252639020	?for i in * do echo $i ; done
252639021	?for i in * do echo $i ; done
252639022	history
[11:59 AM][ttys007 +1][~]
[252639023] mbp13 $ 
```

That junk is not actually in the history file.

```
[12:01 PM][ttys007 +1][~]
[252639024] mbp13 $ od -c /Users/mwilson/.ksh/histfiles/history.ttys007  
0000000  201 001   f   o   r       i       i   n       *       d   o    
0000020    e   c   h   o       $   i       ;       d   o   n   e  \n  \0
0000040  201  \0   h   i   s   t   o   r   y  \n  \0  \0   e   c   h   o
0000060        $   H   I   S   T   F   I   L   E  \n  \0   o   d       -
0000100    c       /   U   s   e   r   s   /   m   w   i   l   s   o   n
0000120    /   .   k   s   h   /   h   i   s   t   f   i   l   e   s   /
0000140    h   i   s   t   o   r   y   .   t   t   y   s   0   0   7  \n
0000160   \0  \0                                                        
0000162

```

If you kill that copy of ksh and reopen one, the junk is gone, so its an in-memory corruption.

```
Version AJMv 93u+m/1.0.0-alpha+c1986c4e 2021-03-06
[12:09 PM][ttys007 +1][~]
[4] mbp13 $ history
1	history
2	echo $HISTFILE
3	od -c /Users/mwilson/.ksh/histfiles/history.ttys007
4	history
[12:09 PM][ttys007 +1][~]
[5] mbp13 $ 
```

Now, that was with a new history file (had to go up to tty007 to get a blank one, I wanted the shell to create a new one).  With an existing populated history file, u+m silently fails to add the line and there appears to be no corruption:

```
[12:07 PM][ttys000 +1][~]
[719] mbp13 $ for i in * do echo $i ; done
-ksh: syntax error: `done' unexpected
[12:07 PM][ttys000 +1][~]
[719] mbp13 $ history -1
718	ls
719	history -1
[12:08 PM][ttys000 +1][~]
[720] mbp13 $ ver
Version AJMv 93u+m/1.0.0-alpha+c1986c4e 2021-03-06
```

ksh 93u+ does not do this, it simply adds the erroneous command line to the history file:

```
Version AJM 93u+ 2012-08-01
[12:12 PM][ttys007 +2][~]
[2] mbp13 $ for i in * do echo $i ; done
/bin/ksh: syntax error: `done' unexpected
[12:12 PM][ttys007 +2][~]
[3] mbp13 $ history
1	/bin/ksh
2	for i in * do echo $i ; done
3	history
[12:12 PM][ttys007 +2][~]
[4] mbp13 $ 
```

**Comment by posguy99** [comment]
It happens on Linux, too.

```
Version AJMv 93u+m/1.0.0-alpha+f8f2c4b6 2021-03-05
[12:20 PM][3 +1][~]
[1] mint $ for i in * do echo $i ; done
ksh: syntax error: `done' unexpected
[12:20 PM][3 +1][~]
[1844518830] mint $ history
1844518815	�for i in * do echo $i ; done
1844518816	�for i in * do echo $i ; done
1844518817	�for i in * do echo $i ; done
1844518818	�for i in * do echo $i ; done
1844518819	�for i in * do echo $i ; done
1844518820	�for i in * do echo $i ; done
1844518821	�for i in * do echo $i ; done
1844518822	�for i in * do echo $i ; done
1844518823	�for i in * do echo $i ; done
1844518824	�for i in * do echo $i ; done
1844518825	�for i in * do echo $i ; done
1844518826	�for i in * do echo $i ; done
1844518827	�for i in * do echo $i ; done
1844518828	�for i in * do echo $i ; done
1844518829	�for i in * do echo $i ; done
1844518830	history
[12:21 PM][3 +1][~]
[1844518831] mint $ 
```

**Comment by posguy99** [comment]
Cleaner u+ demonstration:

```
[12:27 PM][ttys001 +2][~]
[527] mbp13 $ HISTFILE=~/sh_hist
[12:27 PM][ttys001 +2][~]
[1] mbp13 $ for i in * do echo $i ; done
/bin/ksh: syntax error: `done' unexpected
[12:27 PM][ttys001 +2][~]
[2] mbp13 $ history                     
1	for i in * do echo $i ; done
2	history
[12:28 PM][ttys001 +2][~]
[3] mbp13 $ ver
Version AJM 93u+ 2012-08-01
[12:28 PM][ttys001 +2][~]
[4] mbp13 $ 
```

**Comment by posguy99** [comment]
It's syntax errors that aren't being added to the history, if the shell gets to trying to execute it, then its in the history.  I suppose that makes sense so I'm not sure why I'm pointing it out.

```
[12:39 PM][ttys000 +1][~]
[737] mbp13 $ for i in $(seq 1 2) ; do ech $i ; done
-ksh: ech: not found
-ksh: ech: not found
[12:39 PM][ttys000 +1][~]
[738] mbp13 $ history -2
736	ls
737	for i in $(seq 1 2) ; do ech $i ; done
738	history -2
[12:39 PM][ttys000 +1][~]
[739] mbp13 $ 
```

**Comment by McDutchie** [comment]
Thanks for catching that. Now to find the commit that broke it.

**Comment by posguy99** [comment]
That junk line in the history is this:

`\201^Afor i in * do echo $i ; done`

Maybe the first two bytes mean something to someone.

**Comment by JohnoKing** [comment]
This bug was introduced in commit e999f6b16. Syntax errors are added to the history file without odd corruption when that change is reverted.

**Comment by McDutchie** [comment]
> This bug was introduced in commit [e999f6b](https://github.com/ksh93/ksh/commit/e999f6b169f06554dcaa7bfc2ce6c6087d4f7eba).

Confirmed. Good find. Also interesting: I can only reproduce this bug if ksh is compiled with standard malloc.

So, we need to find a better place for that `sfsync()` call. Given the comment, immediately before the `ftruncate()` call would be an obvious place to try...

**Comment by McDutchie** [comment]
Just a datapoint: the e999f6b fix was backported from ksh2020 (see https://github.com/att/ast/issues/61), and sure enough, it also has this bug.

**Comment by McDutchie** [comment]
Patch. No regression tests fail, including those added in e999f6b. And the bug seems to be fixed.

This also changes the style of the call to match historic practice. `NULL` was/is rarely or never used, `0` or [`NIL(sometype)`](https://github.com/ksh93/ksh/blob/c1986c4e1a42bf26230f1138b74ef7c7bc7f173b/src/cmd/ksh93/include/defs.h#L291) is generally used instead. As far as I know, it makes no real difference, but I like to make new code look like it belongs there.
```diff
diff --git a/src/cmd/ksh93/sh/io.c b/src/cmd/ksh93/sh/io.c
index 0f4be76b..4b83a7e3 100644
--- a/src/cmd/ksh93/sh/io.c
+++ b/src/cmd/ksh93/sh/io.c
@@ -1712,11 +1712,6 @@ void	sh_iorestore(Shell_t *shp, int last, int jmpval)
 	register int 	origfd, savefd, fd;
 	int flag = (last&IOSUBSHELL);
 	last &= ~IOSUBSHELL;
-	/*
-	 * There was an issue with truncating files (see 'ftruncate' below) that was caused by
-	 * out-of-sync streams. So, to be safe, sync all streams before restoring file descriptors.
-	 */
-	sfsync(NULL);
 	for (fd = shp->topfd - 1; fd >= last; fd--)
 	{
 		if(!flag && filemap[fd].subshell)
@@ -1740,7 +1735,10 @@ void	sh_iorestore(Shell_t *shp, int last, int jmpval)
 			return;
 		}
 		if(filemap[fd].tname == Empty && shp->exitval==0)
+		{
+			sfsync(NIL(Sfio_t*));
 			ftruncate(origfd,lseek(origfd,0,SEEK_CUR));
+		}
 		else if(filemap[fd].tname)
 			io_usename(filemap[fd].tname,(int*)0,origfd,shp->exitval?2:1);
 		sh_close(origfd);
```

**Comment by JohnoKing** [comment]
The patch works for me.

> Also interesting: I can only reproduce this bug if ksh is compiled with standard malloc.

For some reason I can only get the history corruption bug to occur when I compile ksh with vmalloc, although syntax errors always refuse to enter the history file regardless of which malloc I use.

**Comment by McDutchie** [comment]
Undefined behaviour is undefined, I suppose. How did you manage to find the commit that broke it so quickly?

**Comment by JohnoKing** [comment]
I test the last commit of each month (e.g., if a bug occurs in a July 31 commit but not in the last commit from June 30, the bug was introduced in July). After that I test each commit from the month the bug first appeared until I find the broken one.

**Comment by avih** [comment]
>  I test each commit from the month

`git bisect` could be a very good friend on such case.

**Comment by McDutchie** [comment]
I've taken to keeping a compiled binary of every commit, tagging each file name with the short commit hash. That usually makes it very quick to find the commit that introduced a bug. But so far they've all been compiled with vmalloc and on my system I couldn't reproduce the bug on any of them, presumably because of that.

**Comment by posguy99** [comment]
So that's a Catalina vs Mojave difference?  Mine on macOS are all vmalloc. 

-- 
 Marc Wilson

**Comment by McDutchie** [comment]
It would also be nice to understand a little more about how/why `sh_iorestore()` is involved with the history file, and why syncing the logical streams state with the physical state causes a bug, which seems counterintuitive. `sh_iorestore()` is not called directly from any file within `src/cmd/ksh93/edit`, including the history mechanism.

Maybe these calls in `exfile()` in main.c are the culprit... https://github.com/ksh93/ksh/blob/c1986c4e1a42bf26230f1138b74ef7c7bc7f173b/src/cmd/ksh93/sh/main.c#L418-L420

If `sfsync()` is called directly after `hist_flush()` it would make some sort of sense sense if it doesn't like it if all streams are synched right before `hist_flush()`.

**Comment by McDutchie** [comment]
Which makes me wonder if, after this patch, the bug can still be triggered if we combine a syntax error with a file descriptor that truncates a file.

**Comment by McDutchie** [comment]
> Which makes me wonder if, after this patch, the bug can still be triggered if we combine a syntax error with a file descriptor that truncates a file.

My provisional answer to that is no because the entire line is parsed before anything is executed, even if it contains multiple commands separated by semicolons. So a syntax error would always be thrown before any redirection is executed, which means by the time it gets to that, there are guaranteed to be no syntax errors. This fix should be safe enough.

**Comment by posguy99** [comment]
The patch fixes it for me on macOS Catalina 10.15.7 and Linux Mint.

I don't suppose it matters, but why is the behavior of the bug different depending on whether or not it's the first write to the history file?  If the history file already exists, it just silently fails adding the line to the file.  But if the file doesn't yet exist, and ksh is flushing it out to storage, *then* you get the corruption.

**Comment by McDutchie** [comment]
Beats me. I can't reproduce that behaviour on my end at all.
